src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs

```diff
@@ -74,8 +76,250 @@ public static async ValueTask StoreAsync(string key, OutputCacheEntry value, Tim
 
         using var bufferStream = new MemoryStream();
 
-        JsonSerializer.Serialize(bufferStream, formatterEntry, FormatterEntrySerializerContext.Default.FormatterEntry);
+        Serialize(bufferStream, formatterEntry);
 
         await store.SetAsync(key, bufferStream.ToArray(), value.Tags ?? Array.Empty<string>(), duration, cancellationToken);
     }
+
+    // Format:
+    // Serialization revision:
+    //   7-bit encoded int
+    // Creation date:
+    //   Ticks: 7-bit encoded long
+    //   Offset.TotalMinutes: 7-bit encoded long
+    // Status code:
+    //   7-bit encoded int
+    // Headers:
+    //   Headers count: 7-bit encoded int
+    //   For each header:
+    //     key name byte length: 7-bit encoded int
+    //     UTF-8 encoded key name byte[]
+    //     Values count: 7-bit encoded int
+    //     For each header value:
+    //       data byte length: 7-bit encoded int
+    //       UTF-8 encoded byte[]
+    // Body:
+    //   Segments count: 7-bit encoded int
+    //   For each segment:
+    //     data byte length: 7-bit encoded int
+    //     data byte[]
+    // Tags:
+    //   Tags count: 7-bit encoded int
+    //   For each tag:
+    //     data byte length: 7-bit encoded int
+    //     UTF-8 encoded byte[]
+
+    private static void Serialize(Stream output, FormatterEntry entry)
+    {
+        using var writer = new BinaryWriter(output);
+
+        // Serialization revision:
+        //   7-bit encoded int
+        writer.Write7BitEncodedInt(SerializationRevision);
+
+        // Creation date:
+        //   Ticks: 7-bit encoded long
+        //   Offset.TotalMinutes: 7-bit encoded long
+
+        writer.Write7BitEncodedInt64(entry.Created.Ticks);
+        writer.Write7BitEncodedInt64((long)entry.Created.Offset.TotalMinutes);
+
+        // Status code:
+        //   7-bit encoded int
+        writer.Write7BitEncodedInt(entry.StatusCode);
+
+        // Headers:
+        //   Headers count: 7-bit encoded int
+
+        writer.Write7BitEncodedInt(entry.Headers.Count);
+
+        //   For each header:
+        //     key name byte length: 7-bit encoded int
+        //     UTF-8 encoded key name byte[]
+
+        foreach (var header in entry.Headers)
+        {
+            WriteUTF8String(writer, header.Key);
+
+            if (header.Value == null)
```

Comment:
Should this write 0 still? Otherwise the reader will try to read the length but get the next thing written which could be the next header key length or the body count.

Reply:
Right, also wondering if the value can be null at all in a request. I think I can force it by code, but what would be the value of "header1: ", I would assume `string.Empty` but not null. And assuming we write a length of `0`, should we deserialize it with `""` or `null`.

My preference is to deserialize it as `""` in that case.

Reply:
Well in this case just write 0 for the length.

In the other two cases set value to string.Empty and let `BinaryWriter.Write(string)` handle it.